### PR TITLE
Implement global-workflow on AWS

### LINKED ISSUE #2: Implement global-workflow on AWS
### What new functionality do you need?

Implement global-workflow on AWS

### What are the requirements for the new functionality?

Run ATM forecast only first,
then coupled case,
then DA

### Acceptance Criteria

Run successfully on AWS

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by weihuang-jedi**
PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
 This PR implement global-workflow on AWS.

  Resolves https://github.com/NOAA-EPIC/global-workflow-cloud/issues/2
  
# Type of change
- New feature (adds functionality)

# Change characteristics
- Does this change require a documentation update? YES

# How has this been tested?
Tested on AWS

Example:
- Clone and build on AWS

# Checklist
- [x ] Code changes for GSI-Utils
- [ x] Code changes for UFS-Utils
- [ x] Code Changes for GFS-Utils
- [x ] Code changes for GSI-monitor
- [x ] Code changes for UPP
- [x ] Code changes for UFS-weather-model
- x[ ] Code changes for GSI
- [x ] Code changes for GDAS

### CODE REVIEW

**Comment by weihuang-jedi** [comment]
This is at its early stage, depends on many things. Will mark it ready when things all settled.

File: parm/config/gfs/config.resources
Comment (DavidHuber-NOAA) [inline parm/config/gfs/config.resources:1]: @weihuang-jedi This is probably a little early, but I wanted to point out that when you get to testing cycling, you will want to be careful about how you assign resources for the eobs and anal jobs.  Both of these jobs have bugs in them that cause data to be unassimilated if `npe_node` is not equal to the number of tasks on each node.  See the notes under both jobs in config.resources and issue #2092 for more details.

**Comment by weihuang-jedi** [comment]
Thanks for the heads up.
Yes, I need to be more careful and take care of those situations.
Probably will need a short program to check machine CPUs, and set those
when setup the case.
Let us keep our eyes open on such issue.

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [submodule "sorc/ufs_model.fd"]
   path = sorc/ufs_model.fd
-  url = https://github.com/ufs-community/ufs-weather-model
+  url = https://github.com/ufs-community/ufs-weather-model.git
```
Comment (aerorahul) [inline .gitmodules:3]: please revert.

File: env/AZUREPW.env [hunk 2]
```diff
@@ -0,0 +1,298 @@
+#! /usr/bin/env bash
```
Comment (aerorahul) [inline env/AZUREPW.env:1]: Please remove Azure and Google additions from this PR as this PR is titled "Implemente global-workflow on AWS"

File: parm/config/gfs/config.base [hunk 3]
```diff
@@ -72,9 +72,12 @@ export DO_BUFRSND="@DO_BUFRSND@"         # BUFR sounding products
...
 export DO_NPOESS="@DO_NPOESS@"           # NPOESS products
-export DO_TRACKER="@DO_TRACKER@"         # Hurricane track verification
-export DO_GENESIS="@DO_GENESIS@"         # Cyclone genesis verification
-export DO_GENESIS_FSU="@DO_GENESIS_FSU@" # Cyclone genesis verification (FSU)
+#export DO_TRACKER="@DO_TRACKER@"         # Hurricane track verification
+#export DO_GENESIS="@DO_GENESIS@"         # Cyclone genesis verification
+#export DO_GENESIS_FSU="@DO_GENESIS_FSU@" # Cyclone genesis verification (FSU)
+export DO_TRACKER="NO"         # Hurricane track verification
+export DO_GENESIS="NO"         # Cyclone genesis verification
+export DO_GENESIS_FSU="NO" # Cyclone genesis verification (FSU)
```
Comment (aerorahul) [inline parm/config/gfs/config.base:80]: Please revert.

File: parm/config/gfs/config.resources [hunk 4]
```diff
@@ -59,7 +59,15 @@ case ${machine} in
...
-    npe_node_max=40
+    npe_node_max=36
+    ;;
+  "AZUREPW")
+    export PARTITION_BATCH="compute"
+    npe_node_max=24
+    ;;
+  "GOOGLEPW")
+    export PARTITION_BATCH="compute"
+    npe_node_max=30
```
Comment (aerorahul) [inline parm/config/gfs/config.resources:70]: Please withhold till these platforms are available and the workflow is being ported to them.

File: parm/config/gfs/config.resources [hunk 5]
```diff
@@ -804,7 +812,7 @@ case ${step} in
     ;;
 
   "atmos_products")
-    export wtime_atmos_products="00:15:00"
+    export wtime_atmos_products="01:15:00"
```
Comment (aerorahul) [inline parm/config/gfs/config.resources:815]: This is a significant increase.
Please revert.

File: parm/config/gfs/config.resources [hunk 6]
```diff
@@ -816,7 +824,7 @@ case ${step} in
     ;;
 
   "verfozn")
-    export wtime_verfozn="00:05:00"
+    export wtime_verfozn="00:15:00"
```
Comment (aerorahul) [inline parm/config/gfs/config.resources:827]: Can this be explained?

File: parm/config/gfs/config.stage_ic [hunk 7]
```diff
@@ -9,10 +9,19 @@ source "${EXPDIR}/config.resources" stage_ic
...
+      export CPL_ICEIC="workflow_${CASE}_refactored"
+      export CPL_OCNIC="workflow_${CASE}_refactored"
+      export CPL_WAVIC="workflow_${CASE}_refactored"
+    else
+      echo "PW_CSP: $PW_CSP"
+      export CPL_ATMIC="${CASE}"
+      export CPL_ICEIC="${CASE}"
+      export CPL_OCNIC="${CASE}"
+      export CPL_WAVIC="${CASE}"
+    fi
```
Comment (aerorahul) [inline parm/config/gfs/config.stage_ic:24]: Why is the canned ICs stored in a different directory naming structure on the cloud? This seems unnecessary

File: scripts/exglobal_stage_ic.sh [hunk 8]
```diff
@@ -36,21 +36,30 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
...
       src="${BASE_CPLIC}/${CPL_ATMIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.${ftype}"
       tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}"
-      ${NCP} "${src}" "${tgt}"
-      rc=$?
+      if [ ! -f ${tgt} ]; then
+        ${NCP} "${src}" "${tgt}"
+        rc=$?
+      else
+        rc=0
+      fi
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh:44]: This is not correct.

File: scripts/exglobal_stage_ic.sh [hunk 9]
```diff
@@ -36,21 +36,30 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
...
       ((rc != 0)) && error_message "${src}" "${tgt}" "${rc}"
       err=$((err + rc))
     done
     for ftype in ca_data fv_core.res fv_srf_wnd.res fv_tracer.res phy_data sfc_data; do
       for ((tt = 1; tt <= ntiles; tt++)); do
         src="${BASE_CPLIC}/${CPL_ATMIC:-}/${PDY}${cyc}/${MEMDIR}/atmos/${PDY}.${cyc}0000.${ftype}.tile${tt}.nc"
+       #tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}.tile${tt}.nc"
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh:51]: Please remove commented lines.

File: scripts/exglobal_stage_ic.sh [hunk 10]
```diff
@@ -36,21 +36,30 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
...
             tgt="${COM_ATMOS_RESTART_PREV}/${PDY}.${cyc}0000.${ftype}.tile${tt}.nc"
         fi
-        ${NCP} "${src}" "${tgt}"
-        rc=$?
+        if [ ! -f ${tgt} ]; then
+          ${NCP} "${src}" "${tgt}"
+          rc=$?
+        else
+          rc=0
+        fi
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh:62]: If target is present, it should overwrite.  What happens if the source has changed?

File: scripts/exglobal_stage_ic.sh [hunk 11]
```diff
@@ -59,18 +68,35 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
     # Stage the FV3 cold-start initial conditions to ROTDIR
     YMD=${PDY} HH=${cyc} declare_from_tmpl COM_ATMOS_INPUT
     [[ ! -d "${COM_ATMOS_INPUT}" ]] && mkdir -p "${COM_ATMOS_INPUT}"
+
```
Comment (aerorahul) [inline scripts/exglobal_stage_ic.sh:71]: Why are there changes in this script?  How are they related to porting the workflow to AWS?

File: sorc/com.slurm [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Comment (aerorahul) [inline sorc/com.slurm:1]: Please remove this file.

File: sorc/gdas.cd
Comment (aerorahul) [inline sorc/gdas.cd:1]: Please do not update submodules to which capability is not available on AWS.

File: sorc/setup-case.sh [hunk 13]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (aerorahul) [inline sorc/setup-case.sh:1]: Please delete this file.

File: versions/run.noaacloud.ver [hunk 14]
```diff
@@ -0,0 +1,14 @@
+export stack_intel_ver=2021.3.0
+export stack_impi_ver=2021.3.0
+export spack_env=gsi-addon-env
+#export spack_env=unified-env
```
Comment (aerorahul) [inline versions/run.noaacloud.ver:4]: please remove this line

File: versions/run.noaacloud.ver [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+export ncl_ver=6.6.2
```
Comment (aerorahul) [inline versions/run.noaacloud.ver:6]: NCL is not used on the cloud or in the workflow

File: versions/run.noaacloud.ver [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+export gempak_ver=7.4.2
+
+source "${HOMEgfs:-}/versions/run.spack.ver"
+export spack_mod_path="/work/noaa/epic/role-epic/spack-stack/orion/spack-stack-${spack_stack_ver}/envs/${spack_env}/install/modulefiles/Core"
+export spack_mod_path="/contrib/spack-stack/spack-stack-${spack_stack_ver}/envs/gsi-addon-env/install/modulefiles/Core"
+
+export cdo_ver=2.2.0
+#export R_ver=4.0.2
```
Comment (aerorahul) [inline versions/run.noaacloud.ver:14]: Please remove

File: versions/run.noaacloud.ver [hunk 17]
```diff
@@ -0,0 +1,14 @@
+export stack_intel_ver=2021.3.0
+export stack_impi_ver=2021.3.0
+export spack_env=gsi-addon-env
+#export spack_env=unified-env
+
+export ncl_ver=6.6.2
+export gempak_ver=7.4.2
+
+source "${HOMEgfs:-}/versions/run.spack.ver"
+export spack_mod_path="/work/noaa/epic/role-epic/spack-stack/orion/spack-stack-${spack_stack_ver}/envs/${spack_env}/install/modulefiles/Core"
```
Comment (aerorahul) [inline versions/run.noaacloud.ver:10]: This is an orion line.  Please remove.

File: workflow/hosts.py [hunk 18]
```diff
@@ -15,7 +15,8 @@ class Host:
     """
 
     SUPPORTED_HOSTS = ['HERA', 'ORION', 'JET', 'HERCULES',
-                       'WCOSS2', 'S4', 'CONTAINER', 'AWSPW', 'GAEA']
+                       'WCOSS2', 'S4', 'CONTAINER', 'GAEA',
+                       'AWSPW', 'AZUREPW', 'GOOGLEPW']
```
Comment (aerorahul) [inline workflow/hosts.py:19]: Please remove Azure and Google.

File: workflow/hosts/awspw.yaml [hunk 19]
```diff
@@ -3,20 +3,26 @@ DMPDIR: '/scratch1/NCEPDEV/global/glopara/dump' # TODO: This does not yet exist.
 PACKAGEROOT: '/scratch1/NCEPDEV/global/glopara/nwpara' #TODO: This does not yet exist.
 COMINsyn: '/scratch1/NCEPDEV/global/glopara/com/gfs/prod/syndat' #TODO: This does not yet exist.
 HOMEDIR: '/contrib/${USER}'
-STMP: '/lustre/${USER}/stmp2/'
-PTMP: '/lustre/${USER}/stmp4/'
+STMP: '/lustre/${USER}/stmp/'
+PTMP: '/lustre/${USER}/ptmp/'
 NOSCRUB: ${HOMEDIR}
+NOSCRUB: '/contrib/${USER}/scrub'
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml:9]: Duplicate line

File: workflow/hosts/awspw.yaml [hunk 20]
```diff
@@ -3,20 +3,26 @@ DMPDIR: '/scratch1/NCEPDEV/global/glopara/dump' # TODO: This does not yet exist.
...
+BASE_CPLIC: '/contrib/Wei.Huang/data/ICs'
+CPL_ATMIC: 'C96'
+IC_PREFIX: 'gfs'
+IC_TYPE: 'input'
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml:13]: These are not host specific and should not be in this file.

File: workflow/hosts/awspw.yaml [hunk 21]
```diff
@@ -3,20 +3,26 @@ DMPDIR: '/scratch1/NCEPDEV/global/glopara/dump' # TODO: This does not yet exist.
...
 ACCOUNT: hwufscpldcld
 ACCOUNT_SERVICE: hwufscpldcld
 SCHEDULER: slurm
+ACCOUNT: $USER
```
Comment (aerorahul) [inline workflow/hosts/awspw.yaml:17]: Duplicate

File: workflow/hosts/azurepw.yaml [hunk 22]
```diff
@@ -0,0 +1,24 @@
+BASE_GIT: '/work/noaa/global/glopara/git'
```
Comment (aerorahul) [inline workflow/hosts/azurepw.yaml:1]: Please remove this file from this PR.

Comment (aerorahul) [reply workflow/hosts/azurepw.yaml:1]: These are Orion paths.

File: workflow/hosts/googlepw.yaml [hunk 23]
```diff
@@ -0,0 +1,24 @@
+BASE_GIT: '/work/noaa/global/glopara/git'
```
Comment (aerorahul) [inline workflow/hosts/googlepw.yaml:1]: Please remove this file.
These are orion paths

File: workflow/rocoto/gfs_tasks.py [hunk 24]
```diff
@@ -24,7 +25,32 @@ def stage_ic(self):
...
+
+                if('IC_PREFIX' in cpl_ic.keys()):
+                    cpl_ic_prefix = f"{icdir}/{cpl_ic['IC_PREFIX']}"
+                else:
+                    cpl_ic_prefix = 'gfs'
+                if('IC_TYPE'  in cpl_ic.keys()):
+                    cpl_ic_type = f"{cpl_ic['IC_TYPE']}"
+                else:
+                    cpl_ic_type = 'input'
+                prefix = f"{icdir}/{cpl_ic_prefix}.@Y@m@d/@H/model_data/atmos/{cpl_ic_type}"
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: This should not be needed.

File: workflow/rocoto/rocoto.py [hunk 25]
```diff
@@ -144,8 +145,11 @@ def _create_innermost_task(task_dict: Dict[str, Any]) -> List[str]:
         strings.append(f'\t<partition>{partition}</partition>\n')
     strings.append(f'\t<walltime>{walltime}</walltime>\n')
     strings.append(f'\t<nodes>{nodes}:ppn={ppn}:tpp={threads}</nodes>\n')
-    if memory is not None:
-        strings.append(f'\t<memory>{memory}</memory>\n')
+
+    pw_csp = os.environ.get('PW_CSP')
+    if pw_csp not in ['aws', 'azure', 'google']:
```
Comment (aerorahul) [inline workflow/rocoto/rocoto.py:150]: rocoto.py need not know if this is PW_CSP.
Please outline the issue and we can discuss.

File: workflow/rocoto/tasks.py [hunk 26]
```diff
@@ -214,7 +215,11 @@ def get_resource(self, task_name):
             else:
                 native += ':shared'
         elif scheduler in ['slurm']:
-            native = '--export=NONE'
+            pw_csp = os.environ.get('PW_CSP')
+            if ( pw_csp in ['aws', 'azure', 'google'] ):
+                native = '--export=ALL --exclusive'
+            else:
+                native = '--export=NONE'
```
Comment (aerorahul) [inline workflow/rocoto/tasks.py:222]: This should be handled elsewhere.

File: workflow/rocoto/workflow_xml.py [hunk 27]
```diff
@@ -156,11 +157,21 @@ def _write_crontab(self, crontab_file: str = None, cronint: int = 5) -> None:
...
+             strings = np.append(strings,
+                       [
+                       f'SHELL="/bin/bash"',
+                       f'BASH_ENV="/etc/bashrc"'
+                       ])
+        strings = np.append(strings,
+                  [
+                  f'{cronintstr} {rocotorunstr}',
+                  '#################################################################',
+                  ''])
```
Comment (aerorahul) [inline workflow/rocoto/workflow_xml.py:174]: Can you describe the reason here

**Review by aerorahul — CHANGES_REQUESTED** [review]
Please describe the hard to understand updates in this PR.
Please see specific comments.

**Comment by HenryRWinterbottom** [comment]
@weihuang-jedi Please properly follow the PR template guidelines. See, for example, https://github.com/NOAA-EMC/global-workflow/pull/2688.

**Comment by DavidHuber-NOAA** [comment]
Should this be closed in favor of #2711?

**Comment by aerorahul** [comment]
Closing in favor of #2711

**Pull request closed without merging**